core/src/main/java/com/rudderstack/android/sdk/core/RudderContext.java

```diff
@@ -65,14 +66,19 @@ public class RudderContext {
         // cachedContext is used every time, once initialized
     }
 
-    RudderContext(Application application, String anonymousId, String advertisingId, String deviceToken) {
+    RudderContext(Application application, String anonymousId, String advertisingId, String deviceToken, boolean collectDeviceId) {
         RudderPreferenceManager preferenceManger = RudderPreferenceManager.getInstance(application);
 
         if (TextUtils.isEmpty(anonymousId)) {
-            anonymousId = preferenceManger.getAnonymousId() != null ? preferenceManger.getAnonymousId() : Utils.getDeviceId(application);
-        } else {
-            preferenceManger.saveAnonymousId(anonymousId);
+            anonymousId = preferenceManger.getAnonymousId();
+            if (anonymousId == null && collectDeviceId) {
```

Comment:
@andreasgangso can you expedite this change?

Reply:
This does not have the same effect because getDeviceId() can return null.
The second if check is a fallback if anonymousId is still null, which is either because we configured to not use the deviceId or **getting deviceId failed**.